### PR TITLE
Typecheck options

### PR DESCRIPTION
**Pull request by mhils**
This PR adds strict run-time type checks to the options manager. Implementing this uncovered a whole bunch of inconsistencies regarding which type we specified and what got actually passed around, so this change should improve stability by quite a bit 😄 

This also fixes #1667 and #1668.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -67,7 +67,7 @@
         "cryptography>=1.3, <1.6",
         "cssutils>=1.0.1, <1.1",
         "Flask>=0.10.1, <0.12",
-        "h2>=2.4.1, <3",
+        "h2>=2.4.1, <2.5",
```
Comment (cortesi) [inline setup.py]: This change relevant to the patch?

**Comment by cortesi** [comment]
Excellent. This is a nice improvement.

File: setup.py (hunk 1 above)
Comment (Kriechi) [inline setup.py]: python-hyper is very strong on semver. Using `<3` as upper bound is ok.

File: test/mitmproxy/utils/test_data.py [hunk 2]
```diff
@@ -1,7 +1,8 @@
+import pytest
 from mitmproxy.utils import data
-from . import tutils
 
 
 def test_pkg_data():
     assert data.pkg_data.path("tools/console")
-    tutils.raises("does not exist", data.pkg_data.path, "nonexistent")
```
Comment (Kriechi) [inline test/mitmproxy/utils/test_data.py]: Do we still have a use for our custom `tutils.raises` handler?
Maybe we should ditch it everywhere and move to `pytest.raises`?

File: setup.py (hunk 1 above)
Comment (mhils) [inline setup.py]: I had test failures for 2.5 locally, which was released today. This is to have a clean CI build here. We can bump this again in a separate PR and let CI decide.

Comment (Kriechi) [inline setup.py]: If you can pinpoint the failures, please let @python-hyper know!

File: test/mitmproxy/utils/test_data.py (hunk 2 above)
Comment (cortesi) [inline test/mitmproxy/utils/test_data.py]: I'm -0.001 on this. The string-based exception stuff can be nice, so moving to pytest.raises would be a slight net negative. If we can make an argument that the context handler pytest.raises offers would be useful, I might be +0.001.

**Pull request merged**